### PR TITLE
[AC-1218] Add ability to delete Provider Portals

### PR DESCRIPTION
**Pull request by cyprain-okeke**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Currently, there is no ability to delete Provider Portals. This functionality should be available from the internal Bitwarden Portal.

On the Provider page, there will be an option to delete the Provider at the bottom of the page. A Provider can only be deleted if no organizations are linked to the Provider.

Deleting from

[Bitwarden portal](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1171-18005&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1399%3A12380&show-proto-sidebar=1&mode=design)

[Provider portal - attempting to delete with clients](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1309-18496&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1309%3A18496&show-proto-sidebar=1&mode=design)

[Provider portal - deleting without clients](https://www.figma.com/proto/m8eMk6yeKfOq4AUuyp70ih/Consolidated-Billing?page-id=698%3A10060&type=design&node-id=1710-15961&viewport=-11087%2C521%2C0.25&t=tpHXkqSav41X2FCa-1&scaling=min-zoom&starting-point-node-id=1710%3A15961&show-proto-sidebar=1&mode=design)

## Code changes

- **file.ext:** Description of what was changed and why

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)
Screenshot 2024-04-15 at 11 07 06
Screenshot 2024-04-15 at 11 07 14
Screenshot 2024-04-15 at 11 08 26
Screenshot 2024-04-15 at 11 09 31
Screenshot 2024-04-15 at 16 16 22
Screenshot 2024-04-15 at 11 10 59

### CODE REVIEW

**Comment by trmartin4** [comment]
🤔 We have ongoing work (https://bitwarden.atlassian.net/browse/PM-1943) to migrate the `verify-recover-delete` component to use Tailwind and new reactive form patterns.  If we add a new component here we will need to migrate it as well or introduce the new component using the new patterns.

**Review by ike-kottlowski — COMMENTED** [review]
Everything looks good. But should this be owned by AC? I'm not sure it's within the scope of Auth.

I think this is similar to[ this PR](https://github.com/bitwarden/clients/pull/8445#pullrequestreview-1956828876) which has a comment that poses the same question.

**Review by justindbaur — APPROVED** [review]

**Review by amorask-bitwarden — CHANGES_REQUESTED** [review]
This doesn't look like the correct wording to me. We're deleting the Provider, not the Bitwarden account. Once you complete this action, the account is still active so this is somewhat confusing.

Screenshot 2024-04-15 at 10 58 10 AM

**Comment by cyprain-okeke** [comment]
> This doesn't look like the correct wording to me. We're deleting the Provider, not the Bitwarden account. Once you complete this action, the account is still active so this is somewhat confusing.
> 
> Screenshot 2024-04-15 at 10 58 10 AM

**Review by justindbaur — DISMISSED** [review]

File: apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    // eslint-disable-next-line rxjs-angular/prefer-takeuntil, rxjs/no-async-subscribe
+    this.route.queryParams.pipe(first()).subscribe(async (qParams) => {
+      if (qParams.providerId != null && qParams.token != null && qParams.name != null) {
+        this.providerId = qParams.providerId;
+        this.token = qParams.token;
+        this.name = qParams.name;
+      } else {
+        // FIXME: Verify that this floating promise is intentional. If it is, add an explanatory comment and ensure there is proper error handling.
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.router.navigate(["/"]);
```
Comment (r-tome) [inline apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts:42]:
```suggestion
        await this.router.navigate(["/"]);
```
Await this and you don't have to disable the eslint rule

File: apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      this.formPromise = this.apiService.providerRecoverDeleteToken(this.providerId, request);
+      await this.formPromise;
+      this.platformUtilsService.showToast(
+        "success",
+        this.i18nService.t("providerDeleted"),
+        this.i18nService.t("providerDeletedDesc"),
+      );
+      // FIXME: Verify that this floating promise is intentional. If it is, add an explanatory comment and ensure there is proper error handling.
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.router.navigate(["/"]);
```
Comment (r-tome) [inline apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts:59]: Same thing here, await the promise

**Review by r-tome — CHANGES_REQUESTED** [review]
Looks great, just a couple of things I'd like to see adressed.

File: apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts (hunk 1 above)
Comment (cyprain-okeke) [reply apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts]: resolved

File: apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts (hunk 2 above)
Comment (cyprain-okeke) [reply apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.ts]: resolved

File: libs/common/src/services/api.service.ts [hunk 3]
```diff
@@ -1178,6 +1179,23 @@ export class ApiService implements ApiServiceAbstraction {
     return new ProviderResponse(r);
   }
 
+  providerRecoverDeleteToken(
```
[the remark this answers is not in the document]
Comment (cyprain-okeke) [reply libs/common/src/services/api.service.ts]: this has been move` provider-api.service`

**Review by r-tome — DISMISSED** [review]
Nice work! Thanks for the changes!

**Review by amorask-bitwarden — DISMISSED** [review]

**Review by r-tome — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

File: apps/web/src/locales/en/messages.json [hunk 4]
```diff
@@ -7905,5 +7905,47 @@
   },
   "unassignedItemsBannerSelfHost": {
     "message": "Notice: On May 2, 2024, unassigned organization items will no longer be visible in your All Vaults view across devices and will only be accessible via the Admin Console. Assign these items to a collection from the Admin Console to make them visible."
+  },
+  "deleteProvider": {
+    "message": "Delete provider"
+  },
+  "deleteProviderConfirmation": {
+    "message": "Deleting a provider is permanent and irreversible. Enter your master password to confirm the deletion of the provider and all associated data.",
+    "placeholders": {
```
Comment (JaredSnider-Bitwarden) [inline apps/web/src/locales/en/messages.json:7914]: Nit: the translation doesn't have a placeholder in it? Should the translation have this section?

File: libs/common/src/admin-console/models/request/provider/provider-verify-delete-recover.request.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export class ProviderVerifyDeleteRecoverRequest {
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/admin-console/models/request/provider/provider-verify-delete-recover.request.ts:1]: Nit: the process and components are all called `RecoverDelete` not `DeleteRecover`. Should they match for clarity?

File: bitwarden_license/bit-web/src/app/admin-console/providers/settings/account.component.ts [hunk 6]
```diff
@@ -62,4 +75,64 @@ export class AccountComponent {
...
+      // User cancelled the dialog
+      return false;
+    }
+
+    // User confirmed the dialog so check verification success
+    if (!result.verificationSuccess) {
+      if (result.noAvailableClientVerificationMethods) {
+        // No client-side verification methods are available
+        // Could send user to configure a verification method like PIN or biometrics
+      }
```
Comment (JaredSnider-Bitwarden) [inline bitwarden_license/bit-web/src/app/admin-console/providers/settings/account.component.ts:133]: ⚠️ this block is not possible unless you are passing `{ ...,  clientSideOnlyVerification: true }` in the `UserVerificationDialogOptions`. So, you can remove it.

**Review by JaredSnider-Bitwarden — COMMENTED** [review]
Nice work on this. I skimmed through it quickly and focused on the use of the `UserVerificationDialogComponent` that we discussed. 2 nits and a small warning.

File: apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.html [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+          </p>
+          <p>{{ "deleteProviderRecoverConfirmDesc" | i18n }}</p>
+          <hr />
+          <div class="d-flex">
+            <button
+              type="submit"
+              class="btn btn-danger btn-block btn-submit"
+              [disabled]="form.loading"
+            >
+              <span>{{ "deleteAccount" | i18n }}</span>
```
Comment (JaredSnider-Bitwarden) [inline apps/web/src/app/admin-console/providers/verify-recover-delete-provider.component.html:19]: ❓ Should this button still say delete account?  Or should it say delete provider?

File: libs/common/src/admin-console/models/request/provider/provider-verify-delete-recover.request.ts (hunk 5 above)
Comment (cyprain-okeke) [reply libs/common/src/admin-console/models/request/provider/provider-verify-delete-recover.request.ts:1]: resolved

File: apps/web/src/locales/en/messages.json (hunk 4 above)
Comment (cyprain-okeke) [reply apps/web/src/locales/en/messages.json:7914]: removed the placeholders until we are able to get UserVerificationDialogComponent to work with placeholders

File: bitwarden_license/bit-web/src/app/admin-console/providers/settings/account.component.ts (hunk 6 above)
Comment (cyprain-okeke) [reply bitwarden_license/bit-web/src/app/admin-console/providers/settings/account.component.ts:133]: I have removed this as suggested although this part of the code was changed during the provider delete implementation.

**Review by JaredSnider-Bitwarden — APPROVED** [review]
💯 !

**Review by justindbaur — APPROVED** [review]

**Review by r-tome — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

**Review by amorask-bitwarden — APPROVED** [review]

**Pull request merged**